dev-tools/support-bundle-reader-lib/src/dashboard.rs

```diff
@@ -0,0 +1,454 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Dashboard for inspecting bundles
+
+use crate::bundle_accessor::BoxedFileAccessor;
+use crate::bundle_accessor::InternalApiAccess;
+use crate::bundle_accessor::LocalFileAccess;
+use crate::bundle_accessor::SupportBundleAccessor;
+use crate::index::SupportBundleIndex;
+use anyhow::Context;
+use anyhow::Result;
+use anyhow::bail;
+use bytes::BytesMut;
+use camino::Utf8Path;
+use camino::Utf8PathBuf;
+use crossterm::{
+    event::{self, DisableMouseCapture, EnableMouseCapture, Event, KeyCode},
+    execute,
+    terminal::{
+        EnterAlternateScreen, LeaveAlternateScreen, disable_raw_mode,
+        enable_raw_mode,
+    },
+};
+use futures::TryStreamExt;
+use nexus_client::types::SupportBundleInfo;
+use nexus_client::types::SupportBundleState;
+use omicron_uuid_kinds::GenericUuid;
+use omicron_uuid_kinds::SupportBundleUuid;
+use ratatui::Frame;
+use ratatui::Terminal;
+use ratatui::backend::Backend;
+use ratatui::backend::CrosstermBackend;
+use ratatui::layout::Constraint;
+use ratatui::layout::Layout;
+use ratatui::style::Modifier;
+use ratatui::style::Style;
+use ratatui::widgets::Block;
+use ratatui::widgets::Borders;
+use ratatui::widgets::List;
+use ratatui::widgets::ListState;
+use ratatui::widgets::Paragraph;
+use ratatui::widgets::Wrap;
+use std::io::Write;
+use std::time::Duration;
+use tokio::io::AsyncRead;
+use tokio::io::AsyncReadExt;
+use tokio::io::BufReader;
+
+const BUF_READER_CAPACITY: usize = 1 << 16;
+
+enum FileState<'a> {
+    Open { access: Option<BufReader<BoxedFileAccessor<'a>>>, preview: BytesMut },
+    Closed,
+}
+
+/// A dashboard for inspecting a support bundle contents
+pub struct SupportBundleDashboard<'a> {
+    access: Box<dyn SupportBundleAccessor + 'a>,
+    index: SupportBundleIndex,
+    selected: usize,
+    file: FileState<'a>,
+}
+
+impl<'a> SupportBundleDashboard<'a> {
+    async fn new(access: Box<dyn SupportBundleAccessor + 'a>) -> Result<Self> {
+        let index = access.get_index().await?;
+        if index.files().is_empty() {
+            bail!("No files found in support bundle");
+        }
+        Ok(Self { access, index, selected: 0, file: FileState::Closed })
+    }
+
+    fn index(&self) -> &SupportBundleIndex {
+        &self.index
+    }
+
+    async fn select_up(&mut self, count: usize) -> Result<()> {
+        self.selected = self.selected.saturating_sub(count);
+        if matches!(self.file, FileState::Open { .. }) {
```

Comment:
Hmm maybe I am missing something, but if a file is in the "open" state, why do we proceed to `open_and_buffer`? Unless the naming is weird wouldn't we be opening the file if it's currently closed?

Reply:
I think I understand now that I read some more of the code, but doesn't navigating around cause us to call `open_and_buffer` again on a potentially already opened file that the toggle method has already setup for us? It seems like maybe we want to check if the access is also `None`?

Reply:
This is implemented intentionally, but maybe I can make it more clear in the code:

- When the TUI starts, you see the index of files
- When you move up/down, you're selecting a different file. The files are still closed though
- If you press enter, you'll open a file - this is where we buffer the contents
- If you now move up/down, you're selecting a *different file*. In this case: the file **was** open, but you're picking a new file.

I'll add some comments here to make this more clear, and trigger explicitly when `self.selected` changes